wgpu-hal/src/metal/mod.rs

```diff
@@ -388,8 +390,23 @@ impl crate::Queue<Api> for Queue {
     }
 
     unsafe fn get_timestamp_period(&self) -> f32 {
-        // TODO: This is hard, see https://github.com/gpuweb/gpuweb/issues/1325
-        1.0
+        if !self.shared.private_caps.supports_timestamp_period {
+            return 1.0;
+        }
+
+        use objc::{msg_send, runtime::Object, sel, sel_impl};
+        let timestamps0 = [std::ptr::null_mut::<Object>(); 2];
+        let device = self.shared.device.lock().as_ptr();
+        let () = msg_send![device, sampleTimestamps:&timestamps0[0] gpuTimestamp:&timestamps0[1]];
+        if timestamps0[0] as u64 == 0 || timestamps0[1] as u64 == 0 {
+            return 1.0;
+        }
+
+        let timestamps1 = [std::ptr::null_mut::<Object>(); 2];
+        let () = msg_send![device, sampleTimestamps:&timestamps1[0] gpuTimestamp:&timestamps1[1]];
```

Comment:
shouldn't there be some kind of work between consecutive invocations of the timestamps? or do you think the resolution here is already enough?

Reply:
Apple suggested call it at least twice, once when creating a command buffer and a second time inside the command buffer’s completion handler, but that path can't be implemented simply here, and I don't know if it makes sense to create a command buffer for that?

Reply:
Well, the whole `get_timestamp_period` is not in WebGPU. So we don't have to try implementing it, technically, and could come up with an entirely different API. How about no timestamp period at all? User gets some timestamps, and they can use it for comparison, but it's not really tracking absolute time.

Reply:
Looks like the timestamps API is still in discussion? https://github.com/gpuweb/gpuweb/issues/2642